### PR TITLE
Fix UserLock - user being locked forever.

### PR DESCRIPTION
**Pull request by taOsemc**
# PULL REQUEST

## **Overview**

The verification of a "Locked" user always takes place first, but a user will be set as "Locked" just before the execution of the command to prevent a user from getting blocked forever.

## **Status**

- [X] Typings have been updated or don't need to be.
- [x] This PR have been tested and is ready to be merged.

## **Semantic versioning classification**

- [X] PATCH: This PR fixes a bug, if needed it also references the relevant issue or documentation.

### CODE REVIEW

File: src/Core/Command/CommandUserLock.js [hunk 1]
```diff
@@ -41,18 +39,15 @@ class CommandUserLock {
...
     /**
      * Checks if the user is locked.
-     * Lock the user if not locked.
      * @param {String} userID - The userID
      * @returns {Boolean} Whether the user is locked or not.
-     *
      * @memberof CommandUserLock
      */
     isLocked(userID) {
         const lock = this._usersLocked.has(userID);
```
Comment (Khaaz) [inline src/Core/Command/CommandUserLock.js]: we can directly return this._userLocked.has() here

File: src/Core/Command/Command.js [hunk 2]
```diff
@@ -287,7 +289,10 @@ class Command extends Base {
             }
             return ctx.resolveAsync();
         }
-
+        
+        /** Locking the user to one command instance. */
+        this._userLock.setLock(userID);
```
Comment (Khaaz) [inline src/Core/Command/Command.js]: I'd prefer this to be in the _execute method right now.
We might maybe want another check of isLocked before locking in case there are 2 concurrent run of the command (very edge case but still to take in consideration).

File: src/Core/Command/Command.js [hunk 3]
```diff
@@ -203,6 +203,8 @@ class Command extends Base {
 
         /* Test for userLock */
         const userIsLocked = this._userLock.isLocked(userID);
+
+        /** If user is locked, return */
         if (userIsLocked) {
```
Comment (Khaaz) [inline src/Core/Command/Command.js]: I'd also put the whole this._userLock.isLocked() directly in the if instead of using a variable

**Review by Khaaz — CHANGES_REQUESTED** [review]
I'd also want to point out that this will lock a user to one command execution accross all guilds.

File: src/Core/Command/Command.js [hunk 4]
```diff
@@ -310,6 +309,9 @@ class Command extends Base {
      * @memberof Command
      */
     _execute(env) {
+        /** Locking the user to one command instance. */
+        this._userLock.setLock(env.msg.author.id);
```
Comment (bsian03) [inline src/Core/Command/Command.js]: I'm wondering if it's better to have this in the command dispatcher instead.. cc @Khaaz

File: src/Core/Command/Command.js [hunk 5]
```diff
@@ -287,7 +286,7 @@ class Command extends Base {
             }
             return ctx.resolveAsync();
         }
-
+
```
Comment (bsian03) [inline src/Core/Command/Command.js]:
```suggestion

```
I'm assuming you didn't mean to add the spaces here

File: src/Core/Command/CommandOptions.js [hunk 6]
```diff
@@ -44,7 +44,7 @@ class CommandOptions {
      * @param {Boolean} override.guildOnly - Whether to allow executing this command outside of guilds
      * @param {Boolean} override.hidden - Whether to hide this command from help command (general / subcommands)
      * @param {Number} override.cooldown - Cooldown between each usage of this command for a specific user (in ms)
-     * @prop {Boolean} override.userLock - One user can only run one instance of the command at a time.
+     * @prop {Boolean} override.userLock - One user can only run one instance of the command at a time across all guilds
```
Comment (bsian03) [inline src/Core/Command/CommandOptions.js]:
```suggestion
     * @param {Boolean} override.userLock - One user can only run one instance of the command at a time across all guilds
```
copypast :D

**Review by bsian03 — CHANGES_REQUESTED** [review]
See the things below, also whilst we're here, should we add a public getter for userLock? You'll also need to add the typings for that also (you missed out `_userLock` in your previous PR)

File: src/Core/Command/Command.js (hunk 4 above)
Comment (taOsemc) [reply src/Core/Command/Command.js]: > See the things below, also whilst we're here, should we add a public getter for userLock? You'll also need to add the typings for that also (you missed out `_userLock` in your previous PR)

should be fixed here: https://github.com/Khaaz/AxonCore/pull/120/commits/564092d10c22a78f4e9d1869ea222ca9e8d114c6

**Comment by Khaaz** [comment]
I +1 having a public `shouldLock()` or `shouldUserLock()` to stay consistent with the rest.

**Comment by Khaaz** [comment]
revert last commit, idk how but it introduces additional diff in the PR (94884ec)

File: src/Core/Command/Command.js (hunk 4 above)
Comment (Khaaz) [reply src/Core/Command/Command.js]: We can't really have that in CommandDispatcher since we need the command object anyway. Makes more sense to have the whole userLock in Command directly.

We have 2 possibilities here imo:
- we lock at the same place we check isUserLocked and we unlock everytime we return.
 - we lock before execute

I think it's easier to lock once before execute here. However as I mentionned we COULD have 2 concurrential Command running which would lock it at execute (but the other command would still be running)

File: src/Core/Command/CommandUserLock.js [hunk 7]
```diff
@@ -30,46 +29,33 @@ class CommandUserLock {
 
     /**
      * Returns the status of the userlock option.
-     *
      * @readonly
      * @type {Boolean} Options of userLock.
      * @memberof CommandUserLock
      */
     get userLock() {
```
Comment (Khaaz) [inline src/Core/Command/CommandUserLock.js]: this is probably worth renaming `shouldLock()` as well
For the sake of it and in case we add additional lock in the future, we should probably use `CommandOptions.shouldUserLock()` instead of `CommandOptions.shouldLock()` sorry

Comment (Khaaz) [reply src/Core/Command/CommandUserLock.js]: TLDR: 
`CommandUserLock.shouldLock()`
`CommandOptions.shouldUserLock()`

File: src/Core/Command/Command.js [hunk 8]
```diff
@@ -310,6 +308,9 @@ class Command extends Base {
      * @memberof Command
      */
     _execute(env) {
+        /** Locking the user to one command instance. */
+        this._userLock.setLock(env.msg.author.id);
```
Comment (Khaaz) [inline src/Core/Command/Command.js]:
```suggestion
        this._userLock.setLock(this.library.message.getAuthorID(env.msg) );
```

**Review by Khaaz — APPROVED** [review]
